src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -402,10 +415,37 @@ class LedgerImpl(vm: VM, blockchain: BlockchainImpl, blockchainConfig: Blockchai
         }
       case None => worldState
     }
+  }
+
+  /**
+    * EIP161 - State trie clearing
+    * Delete all accounts that have been touched (involved in any potentially state-changing operation) during transaction execution.
+    *
+    * All potentially state-changing operation are:
+    * Account is the target or refund of a SUICIDE operation for zero or more value;
+    * Account is the source or destination of a CALL operation or message-call transaction transferring zero or more value;
+    * Account is the source or newly-creation of a CREATE operation or contract-creation transaction endowing zero or more value;
+    * as the block author ("miner") it is recipient of block-rewards or transaction-fees of zero or more.
+    *
+    * Deletion of touched account should be executed immediately following the execution of the suicide list
+    *
+    * @param worldStateProxy world after execution of all potentially state-changing operations
+    * @return a worldState equal worldStateProxy except that the accounts touched during exection are deleted and touched
+    *         Set is cleared
+    */
+  private[ledger] def deleteTouchedAccounts(worldStateProxy: InMemoryWorldStateProxy): InMemoryWorldStateProxy = {
+    def deleteEmptyAccounts(addressesToDelete: Set[Address]) =
+      addressesToDelete.foldLeft(worldStateProxy){ case (world, address) => deleteEmptyAccount(world, address) }
 
+    def deleteEmptyAccount(world: InMemoryWorldStateProxy, address: Address) = {
+      if (world.getAccount(address).exists(_.isEmpty))
+        world.deleteAccount(address)
+      else
+        world
```

Comment:
Do we need to check for existence? Isn't `deleteAccount` idempotent when account does not exist?

Reply:
This check is for the sake of `_.isEmpty` method as it is called on retrieved Account.  And we cant call here `getGuaranteedAccount` as Account could be earlier deleted in suicide list (as it is executed before touched accounts )

Reply:
Ok, I didn't read too carefully, but my point here was that this function seems more verbose than it needs to be. What about something like:

```scala
val addressesToDelete = world.touchedAccounts.filter(addr => world.getAccount(addr).exists(_.isEmpty))
addressesToDelete.foldLeft(world)(_ deleteAccount _).clearTouchedAccounts
```

Reply:
I am not convinced, we are saving some space definitely, but we lose readability ( i can definitely imagine myself staring in some time at `_ deleteAccount _` and struggling to decipher it) and we create unnecessary transition collection of `addressesToDelete`.

Reply:
I'm fine with explicitly named params. Also `clearTouchedAccounts` can be a separate line.

The main point here is that this logic consists of 3 easy steps:

1. determine the addresses to delete
2. delete them
3. clear touched set

IMO, using closures does not make it so obvious (it's also slightly less efficient).

As for efficiency, yes, we could do away with the set transformations by checking account inside `foldLeft`'s inner body, but do we really have a lot to gain here?

Reply:
I don't have big this address set could be but doing some unnecessary work when we can easily avoid it is not good option I think. As for readability it is I think really subjective.

Maybe we can meet in the middle with something like:
`    world.touchedAccounts
      .foldLeft(world){ case (actualWorld, address) => deleteEmptyAccount(actualWorld, address) }
      .clearTouchedAccounts`

In there we abandon one closure, see that we are operating on touched accounts and deleting only empty accounts from that set, and clearing set at the end.

Reply:
I would be fine with that, but I still think my proposition is better. At any rate we don't need the outer closure, cause that can simply be a `val`. I can see some usefulness of the other one - `deleteEmptyAccount`, but I would still skip cause the body of `foldLeft` would be small enough.

Regarding efficiency, all the work of checking accounts and deleting accounts has to be done anyway, and creating the intermediate collections seems small enough compared to that. But if we're concerned, we can still avoid that by using a view: `touchedAccounts.view.flatMap(..).filter(..)`

As for readability: the problem can be formulated in English in these 2 ways:

  - find empty accounts in touched accounts
  - delete them

vs

  - go through all the touched accounts, and if they're empty, delete them

I know the difference seems negligible with such a small problem, but which one sounds clearer?

Reply:
I don't like this `.view` idea as it brings another complication in code which supposes to do a simple thing and can be implemented without it.
As for English readability, the two-step style has its merits (two clearly separated steps), the second is really close and trumps it a little with performance without additional combinators.
Also, I think `deleteEmptyAccount` closure is necessary imo, as this `if .. else` combined with``{ case (actualWorld, address)` just look bad.

Summarising;
I would stay with current approach, but if you insist I can go your way as problem at hand is not some big data set transformation and I would like to close this pr before more merge conflicts arise

Reply:
Both are reasonable. I'm slightly in favor of the 2 liner, but lets go with the authors approach in this case. (Now fun's over - back to work ;) )

Reply:
> I don't like this .view idea as it brings another complication in code

How is:

```scala
val addressesToDelete = world.touchedAccounts.view.filter(addr => world.getAccount(addr).exists(_.isEmpty))
```

or 

```scala
val addressesToDelete = world.touchedAccounts.withFilter(addr => world.getAccount(addr).exists(_.isEmpty))
```

more complicated than:

```scala
val addressesToDelete = world.touchedAccounts.filter(addr => world.getAccount(addr).exists(_.isEmpty))
```
?

Usage of those methods simply implies lazy evaluation. Should we avoid the keyword `lazy` because that increases complexity?

Anyway, all my points were made. If you want to go with the closures, that's good enough for me. Just please get rid off that outer one, like you proposed above.

Also, consider these pointers:
- .foldLeft(z){ (a, x) => ... } - no need for the `case` keyword
- you can write `touchedAccounts.foldLeft(world)(deleteEmptyAccount)`

I'm ready to approve this one, pending your decision here.

Reply:
> Usage of those methods simply implies lazy evaluation. Should we avoid the keyword lazy because that increases complexity <

I am not saying to avoid it, but to use it with care e.g not use `lazy val` when `val` will suffice. I just think that lazy evaluation in collections (e.g `Streams`, `withFilter`)is great when we have many operations stacked up onto large collections. In the current situation, it seems overkill as we have just one condition and operation.

>you can write touchedAccounts.foldLeft(world)(deleteEmptyAccount)

👍